R/chat-tools-content.R

```diff
@@ -0,0 +1,72 @@
+# Very few providers support anything other than text results; fortunately we
+# can fake them them by unrolling the tool result to a forward pointer to
+# other user content items.
+#
+# ContentToolResult(value = ContentImageInline(...))
+# ->
+# ContentToolResult("See <content-tool> below")
+# ContentText("<content-tool>")
+# ContentImageInline(...)
+# ContentText("</content-tool>")
+
+turn_contents_expand <- function(turn) {
+  if (length(turn@contents) == 0) {
+    # Early return to avoid unlist(list()) yielding NULL
+    return(turn)
+  }
+
+  contents <- map(turn@contents, expand_content_if_needed)
+  turn@contents <- unlist(contents, recursive = FALSE)
+  turn
+}
+
+expand_content_if_needed <- function(content) {
+  if (!is_tool_result(content)) {
+    return(list(content))
+  }
+
+  request <- content@request
+  value <- content@value
+
+  if (S7_inherits(value, Content)) {
+    expand_tool_value(request, value)
+  } else if (is.list(value)) {
+    if (all(map_lgl(value, \(x) S7_inherits(x, Content)))) {
+      expand_tool_values(request, value)
+    } else {
+      list(content)
+    }
+  } else {
+    list(content)
+  }
+}
+
+expand_tool_value <- function(request, value) {
+  open <- sprintf('<tool-content call-id="%s">', request@id)
```

Comment:
Good call adding in the call id!

There's a lot of overlap between `expand_tool_value` and `expand_tool_values`. I'd be tempted to merge them with a small if/else for single vs list values (while keeping the value checking logic in `expand_content_if_needed()`). It's a small thing though and seems like the kind of thing where you'd have a style preference.